### PR TITLE
commit updates #836

### PR DESCRIPTION
**Pull request by salomon-j**

### CODE REVIEW

File: grails-app/services/au/org/ala/ecodata/MetadataService.groovy [hunk 1]
```diff
@@ -135,8 +135,32 @@ class MetadataService {
...
+    }
+
+    /**
+     * Returns a Map of activity types grouped by category.
+     * @param programId Most of the legacy projects are updated recently to insert programId (see github #2741)
+     * @return a Map, key: String, value: List of name, description for each activity in the category
+     */
+    Map activitiesListByProgramId(String programId) {
+        Map config = Program.findByProgramId(programId)?.config
+        List<String> activities = config.activities.collect{it.name}
```
Comment (chrisala) [inline grails-app/services/au/org/ala/ecodata/MetadataService.groovy:148]: Need a null safe navigator here.  It'd definitely possible the config has no "activities" defined.

File: grails-app/services/au/org/ala/ecodata/MetadataService.groovy [hunk 2]
```diff
@@ -135,8 +135,32 @@ class MetadataService {
...
 
+        List<ActivityForm> forms = activityFormService.search(category:activities)
+        List results = forms.collect {[name:it.name, category:it.category, type:it.type, description:it.description, status:it.status]}
+        results = results.findAll {!it.status || it.status == ACTIVE}
```
Comment (chrisala) [inline grails-app/services/au/org/ala/ecodata/MetadataService.groovy:152]: Why is this status based filtering necessary?  If there is bad data in the database (for the status field) we should fix that rather than do this workaround.

File: grails-app/services/au/org/ala/ecodata/MetadataService.groovy [hunk 3]
```diff
@@ -135,8 +135,32 @@ class MetadataService {
...
+
```
Comment (chrisala) [inline grails-app/services/au/org/ala/ecodata/MetadataService.groovy:153]: As implemented, I think this will result in duplicate activity forms where there is more than one form version defined.
For activities used by the original MERIT project template, we'll probably get away with that (because they all have only one version) but it should be fixed or we should add a new issue to track this.

**Review by chrisala — CHANGES_REQUESTED** [review]
Please add a unit test for this.

File: grails-app/services/au/org/ala/ecodata/MetadataService.groovy (hunk 3 above)
Comment (salomon-j) [reply grails-app/services/au/org/ala/ecodata/MetadataService.groovy:155]: Hi Chris, I've pushed my updates, please have a look when you have time.

File: grails-app/services/au/org/ala/ecodata/MetadataService.groovy [hunk 4]
```diff
@@ -135,8 +135,34 @@ class MetadataService {
...
+    }
+
+    /**
+     * Returns a Map of activity types grouped by category.
+     * @param programId Most of the legacy projects are updated recently to insert programId (see github #2741)
+     * @return a Map, key: String, value: List of name, description for each activity in the category
+     */
+    Map activitiesListByProgramId(String programId) {
+        Map config = Program.findByProgramId(programId)?.config
+        List<String> activities = config?.activities.collect{it.name}
```
Comment (chrisala) [inline grails-app/services/au/org/ala/ecodata/MetadataService.groovy:148]: This will still throw an NPE if the config has no activities attached.  Please include in your "missing config" unit test a scenario includes a Program with a non-null config object with no "activities" key.

File: grails-app/services/au/org/ala/ecodata/MetadataService.groovy [hunk 5]
```diff
@@ -135,8 +135,34 @@ class MetadataService {
...
 
+        List forms = activityFormService.search(category:activities)
+        forms = forms.findAll{it.publicationStatus == PublicationStatus.PUBLISHED}
```
Comment (chrisala) [inline grails-app/services/au/org/ala/ecodata/MetadataService.groovy:151]: Searching for published forms is probably a good idea here, but you should pass it as a criteria rather than doing post-filtering.

**Review by chrisala — CHANGES_REQUESTED** [review]

File: grails-app/services/au/org/ala/ecodata/MetadataService.groovy (hunk 4 above)
Comment (salomon-j) [reply grails-app/services/au/org/ala/ecodata/MetadataService.groovy:148]: i've added this test method, should this be ok?

Screenshot 2023-05-16 at 10 53 14 am
